### PR TITLE
Support filtering data in Auto Compaction

### PR DESCRIPTION
**Pull request by maytasm**
Support filtering data in Auto Compaction

### Description

This PR adds `filter` field via the `transformSpec` to manual compaction and auto compaction.
By setting `filter`, the user will be able to drop data matching the `filter` from the datasource.

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/client/indexing/ClientCompactionTaskTransformSpec.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  @JsonCreator
+  public ClientCompactionTaskTransformSpec(
+      @JsonProperty("filter") final DimFilter filter
+  )
+  {
+    this.filter = filter;
+  }
+
+  @Nullable
+  public DimFilter getFilter()
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/client/indexing/ClientCompactionTaskTransformSpec.java:49]: is it missing `JsonProperty` annotation?

File: server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java [hunk 2]
```diff
@@ -364,6 +365,16 @@ private CoordinatorStats doRun(
...
 
+        // Create transformSpec to send to compaction task
+        ClientCompactionTaskTransformSpec transformSpec;
+        if (config.getTransformSpec() != null) {
+          transformSpec = new ClientCompactionTaskTransformSpec(
+              config.getTransformSpec().getFilter()
+          );
+        } else {
+          transformSpec = null;
+        }
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java:376]:
```suggestion
        ClientCompactionTaskTransformSpec transformSpec = null;
        if (config.getTransformSpec() != null) {
          transformSpec = new ClientCompactionTaskTransformSpec(
              config.getTransformSpec().getFilter()
          );
        } 
```

Comment (abhishekagarwal87) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java:376]: nit - could be simplified.

**Review by abhishekagarwal87 — COMMENTED** [review]
is the PR going to add more changes? One thing I found interesting is that we are re-compacting the data if the filter changes. But the data we are compacting has already passed through a filter. Those records are gone and are not going to come back. Say if I first do a filter on `dim = a` and then change the filter to `dim = b`. It might be nice to add a warning in the UI for this when we eventually add those changes.

File: server/src/main/java/org/apache/druid/client/indexing/ClientCompactionTaskTransformSpec.java (hunk 1 above)
Comment (maytasm) [reply server/src/main/java/org/apache/druid/client/indexing/ClientCompactionTaskTransformSpec.java:53]: Yes. Fixed

File: server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java (hunk 2 above)
Comment (maytasm) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java:376]: Done

**Comment by maytasm** [comment]
> is the PR going to add more changes? One thing I found interesting is that we are re-compacting the data if the filter changes. But the data we are compacting has already passed through a filter. Those records are gone and are not going to come back. Say if I first do a filter on `dim = a` and then change the filter to `dim = b`. It might be nice to add a warning in the UI for this when we eventually add those changes.

I have finish adding all changes to this PR (implementation, unit tests, docs, integration tests). Your understanding is correct. That is now true for other auto compaction config that can remove data. For example, if you remove one of your dimension using auto compaction then add that dimension back later, the removed dimension will not come back (although that dimension in future data will not be removed). I'll work on clarifying this in the docs to make sure that user are well aware of this new behavior in a separate PR.

File: core/src/test/java/org/apache/druid/timeline/DataSegmentTest.java [hunk 3]
```diff
@@ -145,7 +146,7 @@ public void testV1Serialization() throws Exception
     Assert.assertEquals(ImmutableMap.of("type", "numbered", "partitionNum", 3, "partitions", 0), objectMap.get("shardSpec"));
     Assert.assertEquals(TEST_VERSION, objectMap.get("binaryVersion"));
     Assert.assertEquals(1, objectMap.get("size"));
-    Assert.assertEquals(4, ((Map) objectMap.get("lastCompactionState")).size());
+    Assert.assertEquals(5, ((Map) objectMap.get("lastCompactionState")).size());
```
Comment (suneet-s) [inline core/src/test/java/org/apache/druid/timeline/DataSegmentTest.java:149]: this is a test for "V1" Shouldn't this test remain without the `filter` being part of the CompactionState?

Comment (suneet-s) [reply core/src/test/java/org/apache/druid/timeline/DataSegmentTest.java:149]: Scratch that. It's v1 of DataSegment. This seems fine.

File: server/src/test/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstPolicyTest.java [hunk 4]
```diff
@@ -1144,7 +1149,123 @@ public void testIteratorReturnsSegmentsAsSegmentsWasCompactedAndHaveDifferentDim
...
+        Collections.emptyMap()
+    );
+    // No more
+    Assert.assertFalse(iterator.hasNext());
+  }
+
+  @Test
+  public void testIteratorReturnsSegmentsAsSegmentsWasCompactedAndHaveDifferentFilter() throws Exception
+  {
+    NullHandling.initializeForTests();
```
Comment (suneet-s) [inline server/src/test/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstPolicyTest.java:1165]: nit: I think this should be done at the test class level.

**Review by suneet-s — APPROVED** [review]
Overall LGTM.

2 high level comments since you are making a lot of changes in this area:
1. It would be good to have some sort of integration test with segments in different compaction states so that we can simulate an upgrade where users have segments with different compaction states and validate that we don't re-compact segments each time we add a new field here.
2. It would be good to add a warning in the docs that filtering data that matches an entire segment may not work as expected since Druid doesn't handle "empty segments" too well today.

All my comments do not need to block this PR, and can be addressed in future changes.

**Pull request merged**